stunnel service: allow specifying listening ip

martinetd opened the pull request:
##### Motivation for this change

- server side accept in the config syntax is actually [host:]port, it is necessary to specify the listening ip for e.g. ipv6 where one would set `:::port` as a string.
- examples incorrectly had 'enable' set, the option is not defined and reproducing would error out

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

martinetd committed:
stunnel service: fix servers example

examples incorrectly had 'enable' set, the option is not defined
and reproducing would error out

martinetd committed:
stunnel: make servers accept more lenient

stunnel config's accept syntax is [host:]port -- this is required to e.g. listen on ipv6
where one would set :::port

martinetd commented:
sneaked in an actual change, so updated title/original message and Ccing maintainer: Cc @lschuermann

thanks!

lschuermann commented:
Thanks for the contribution, sorry for the delay! The change looks fine to me, works and is in line with the documentation:
```
accept = [HOST:]PORT
   accept connections on specified address
   If no host specified, defaults to all IPv4 addresses for the local host.
   To listen on all IPv6 addresses use:
       accept = :::PORT
```

Maybe we want to also document this behavior? It's quite unconventional to not put IPv6 addresses in square brackets (which will break). Otherwise, this looks fine.

martinetd commented:
> Maybe we want to also document this behavior? It's quite unconventional to not put IPv6 addresses in square brackets (which will break). Otherwise, this looks fine.

Good point, I've updated the description to the following (and rebased to master while I was at it)
```
On which [host:]port stunnel should listen for incoming TLS connections.
Note that unlike other softwares stunnel ipv6 address need no brackets,
so to listen on all IPv6 addresses on port 1234 one would use ':::1234'.
```

Thanks!

lschuermann commented on nixos/modules/services/networking/stunnel.nix:
```diff
@@ -16,8 +16,12 @@ let
+	  On which [host:]port stunnel should listen for incoming TLS connections.
+	  Note that unlike other softwares stunnel ipv6 address need no brackets,
+	  so to listen on all IPv6 addresses on port 1234 one would use ':::1234'.
+	'';
```
Thanks for the documentation, this looks great!

Would you mind fixing the indentation here? Essentially the multi-line string start token should be on the same level as the end token. The white space in front should then be removed automatically.

martinetd replied:
Are you sure? I see the indentation I used all over the place (`git grep -A 5 "description = ''"`)
I had disabled documentation building but will turn it back on to check, will repost in a few minutes

lschuermann replied:
I'm quite certain Nix will strip the whitespace. It might be nitpicking, though I think this identation style is more common (and more pleasing to look at):

```
$ cat test.nix 
let 
  a = ''
    test
  '';

in
  a

$ nix repl
Welcome to Nix version 2.3.6. Type :? for help.

nix-repl> import ./test.nix
"test\n"
```

lschuermann replied:
Oh, I just now realize you've been using tabs, which render for me as 2 spaces, though probably will render as 8 spaces for you. :) Given that the rest of the file uses spaces, that might be more reliable.

lschuermann commented on nixos/modules/services/networking/stunnel.nix:
```diff
@@ -16,8 +16,12 @@ let
-        description = "On which port stunnel should listen for incoming TLS connections.";
+        type = types.either types.str types.int;
+        description = ''
+	  On which [host:]port stunnel should listen for incoming TLS connections.
+	  Note that unlike other softwares stunnel ipv6 address need no brackets,
+	  so to listen on all IPv6 addresses on port 1234 one would use ':::1234'.
+	'';
```
Something like this should work (sorry, if this might not apply, GitHub's UI is not very intuitive for this)
```suggestion
          On which [host:]port stunnel should listen for incoming TLS connections.
          Note that unlike other softwares stunnel ipv6 address need no brackets,
          so to listen on all IPv6 addresses on port 1234 one would use ':::1234'.
        '';
```

martinetd reviewed (commented):
Ah! Sorry, hadn't noticed because it looked good here in vim -- you'd think by 2020 editors could automatically expand tabs based on context... :/
Force pushed what amounts to your suggestion, thanks for insisting!

lschuermann commented:
Pinging @JohnAZoidberg as a maintainer.

JohnAZoidberg reviewed (approved):
Great, thanks!

> you'd think by 2020 editors could automatically expand tabs based on context... :/

We [don't use tabs](https://github.com/NixOS/nixpkgs/blob/master/doc/contributing/coding-conventions.xml#L16) at all ;)

You can use `:set expandtab` to expand the tabs always to the same amount.

martinetd commented:
> We [don't use tabs](https://github.com/NixOS/nixpkgs/blob/master/doc/contributing/coding-conventions.xml#L16) at all ;)
> 
> You can use `:set expandtab` to expand the tabs always to the same amount.

I know, I really need to take 5 minutes to properly add per project/directory settings... :)

Anyway, thanks for merging!
